### DISCUSSION TITLE
Unparsed Response Object For BatchGet

### DISCUSSION THREAD

**Discussion by julian-dotcom** [sensedeep/dynamodb-onetable#536] [Q&A] [answered]
I have an array of `userIds[]` and want to send a batch request to my One Table DynamoDB to get all corresponding users.

Thus far, this has been my (suboptimal) approach, but it requires sending hundreds of API requests
`const items = await Promise.all(userIds.map((id) => this.model.get({ id })))`

Much more preferable would be to use the `batchGet` method, as seen in the documentation:
https://doc.onetable.io/api/table/methods/#batchget

```javascript
let batch = {}
await Account.get({id: accountId}, {batch})
await User.get({id: userId}, {batch})
let results = await table.batchGet(batch)
```

However, the response is completely untyped and looks like a response straight from DynamoDB.

I do not want to write a type guard, as the schema is quite complicated, and a type guard would be difficult to maintain.

Does the lib provide some elegant mechanism to type check such responses?

**Comment by mobsense** [comment]
Add 

```
{
    parse: true
}
```
as params to the batchGet. Can also use hidden:true to suppress hidden fields.

**Reply by julian-dotcom** [reply]
Thanks for the swift response. Unfortunately, when I set `{ parse: true }`, the response is an empty array, similar to this [post](https://github.com/sensedeep/dynamodb-onetable/issues/338). 

I also made sure that all my records have a `_type` property, as by this [comment](https://github.com/sensedeep/dynamodb-onetable/issues/338#issuecomment-2116327760).

Please note that when I set `{ parse: false }`, I do get the desire records, meaning that there is in fact data to be returned.

**Do you have any other suggestions that might lead to the failure of the processing?**

**Comment by mobsense** [comment]
Two suggestions:

add {log: true} so you can see the actual command being executed.

Create a debug.ts that replicates the issue so can can investigate.

**Reply by julian-dotcom** [reply]
Here is the response object, when `{ log: true }`

Does the problem lie with the fact that `"model": "_Generic"`?
```
OneTable "batchGet" "_Generic" {
     "trace": {
         "model": "_Generic",
         "cmd": {
             "RequestItems": {
                 "table-name": {
                     "Keys": [
                         {
                             "pk": {
                                 "S": "User:d191804b-eb8b-45a8-bc23-4a3d4231acdd"
                             },
                           "sk": {
                                 "S": "User:"
                             }
                         }
                     ],
                     "ConsistentRead": false
                 }
             }
         },
         "op": "batchGet",
         "properties": {}
     }
 }
```

**Reply by julian-dotcom** [reply]
To clarify, when `{ parse: false }`, I get the desire response, just unparsed. 

So the problem does not lie with the DynamoDB request, but rather with the parsing.

**Reply by julian-dotcom** [reply, marked as answer]
I figured out the issue, I wasn't passing the model to the table:

```javascript
    return new Table({
      (...)
      models: { User: UserSchema.fields }, // IMPORTANT
    });
  },
  ```

**Reply by julian-dotcom** [reply]
There is still one remaining problem.

While the response is successfully parsed, it is not type safe.

For this `const items = await this.table.batchGet(batch, { parse: true });`, this is the type of items: `const items: {}[]`

<img width="519" alt="Screenshot 2024-08-12 at 17 15 56" src="https://github.com/user-attachments/assets/9ab19b4e-8b6b-4977-ba53-58eeda5ff449">

**Comment by mobsense** [comment]
Great you figured it out. Yes, you need a schema to be able to parse the response.

You wont get a batch response typed -- as the return data types are unknown at dev time.
